### ISSUE TITLE
java.util.ConcurrentModificationException in branch andreistefanescu:kore1

### ISSUE THREAD

**Issue by yilongli** [kframework/k-legacy#1658]
We see several java.util.ConcurrentModificationException in the ongoing pull request: https://github.com/kframework/k/pull/1639.

RV-Predict is able to pinpoint the root cause in one run. Here is the race report:

```
Data race on field java.util.HashMap.$state: {{{
    Concurrent read in thread T55 (locks held: {})
 ---->  at com.rits.cloning.FastClonerHashMap.clone(FastClonerHashMap.java:17)
        at com.rits.cloning.Cloner.fastClone(Cloner.java:114)
        at org.kframework.backend.java.symbolic.DeepCloner$FastTermCloner.fastClone(DeepCloner.java:47)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:416)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneArray(Cloner.java:474)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:428)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneArray(Cloner.java:474)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:428)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneArray(Cloner.java:474)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:428)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.cloneObject(Cloner.java:453)
        at com.rits.cloning.Cloner.cloneInternal(Cloner.java:431)
        at com.rits.cloning.Cloner.deepClone(Cloner.java:301)
        at org.kframework.backend.java.symbolic.DeepCloner.clone(DeepCloner.java:30)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.construct(KAbstractRewriteMachine.java:231)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.rewrite(KAbstractRewriteMachine.java:98)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.rewrite(KAbstractRewriteMachine.java:64)
        at org.kframework.backend.java.symbolic.PatternMatchRewriter.computeRewriteStep(PatternMatchRewriter.java:231)
        at org.kframework.backend.java.symbolic.PatternMatchRewriter.rewrite(PatternMatchRewriter.java:92)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.patternMatcherRewriteRun(JavaSymbolicExecutor.java:108)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.javaRewriteEngineRun(JavaSymbolicExecutor.java:135)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.run(JavaSymbolicExecutor.java:74)
        at org.kframework.krun.tools.Executor$Tool.execute(Executor.java:169)
        at org.kframework.krun.tools.Executor$Tool.run(Executor.java:125)
        at org.kframework.transformation.TransformationCompositionProvider$2.run(TransformationCompositionProvider.java:120)
        at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:131)
        at org.kframework.main.FrontEnd.main(FrontEnd.java:52)
        at org.kframework.main.Main.runApplication(Main.java:109)
        at org.kframework.kserver.KServerFrontEnd.run(KServerFrontEnd.java:119)
        at org.kframework.main.Main.nailMain(Main.java:70)
        at com.martiansoftware.nailgun.NGSession.run(NGSession.java:325)
    T55 is created by T13
        at com.martiansoftware.nailgun.NGSessionPool.take(NGSessionPool.java:85)

    Concurrent write in thread T59 (locks held: {Monitor@5aaa5e9e})
 ---->  at org.kframework.backend.java.kil.Definition.putSortCacheValue(Definition.java:436)
        - locked Monitor@5aaa5e9e at org.kframework.backend.java.kil.Definition.putSortCacheValue(Definition.java:435) 
        at org.kframework.backend.java.kil.KItem.computeSort(KItem.java:241)
        at org.kframework.backend.java.kil.KItem.sort(KItem.java:651)
        at org.kframework.backend.java.symbolic.PatternMatcher.add(PatternMatcher.java:210)
        at org.kframework.backend.java.symbolic.PatternMatcher.stop(PatternMatcher.java:167)
        at org.kframework.backend.java.symbolic.AbstractUnifier.unify(AbstractUnifier.java:144)
        at org.kframework.backend.java.symbolic.PatternMatcher.patternMatch(PatternMatcher.java:145)
        at org.kframework.backend.java.symbolic.PatternMatcher.patternMatch(PatternMatcher.java:136)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.match(KAbstractRewriteMachine.java:258)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.match(KAbstractRewriteMachine.java:323)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.match(KAbstractRewriteMachine.java:299)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.match(KAbstractRewriteMachine.java:323)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.match(KAbstractRewriteMachine.java:323)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.rewrite(KAbstractRewriteMachine.java:68)
        at org.kframework.backend.java.rewritemachine.KAbstractRewriteMachine.rewrite(KAbstractRewriteMachine.java:64)
        at org.kframework.backend.java.symbolic.PatternMatchRewriter.computeRewriteStep(PatternMatchRewriter.java:231)
        at org.kframework.backend.java.symbolic.PatternMatchRewriter.rewrite(PatternMatchRewriter.java:92)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.patternMatcherRewriteRun(JavaSymbolicExecutor.java:108)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.javaRewriteEngineRun(JavaSymbolicExecutor.java:135)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.run(JavaSymbolicExecutor.java:74)
        at org.kframework.krun.tools.Executor$Tool.execute(Executor.java:169)
        at org.kframework.krun.tools.Executor$Tool.run(Executor.java:125)
        at org.kframework.transformation.TransformationCompositionProvider$2.run(TransformationCompositionProvider.java:120)
        at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:131)
        at org.kframework.main.FrontEnd.main(FrontEnd.java:52)
        at org.kframework.main.Main.runApplication(Main.java:109)
        at org.kframework.kserver.KServerFrontEnd.run(KServerFrontEnd.java:119)
        at org.kframework.main.Main.nailMain(Main.java:70)
        at com.martiansoftware.nailgun.NGSession.run(NGSession.java:325)
    T59 is created by T13
        at com.martiansoftware.nailgun.NGSessionPool.take(NGSessionPool.java:85)
}}}
```

**Comment by dwightguth** [comment]
Looks like you need to tell the cloner not to clone Definition.

**Comment by andreistefanescu** [comment]
Fixed in #1639.
